distribution/pull_v2.go

```diff
@@ -403,10 +403,20 @@ func (p *v2Puller) pullV2Tag(ctx context.Context, ref reference.Named) (tagUpdat
 		if err = p.config.ReferenceStore.AddDigest(canonical, imageID, true); err != nil {
 			return false, err
 		}
-	} else if err = p.config.ReferenceStore.AddTag(ref, imageID, true); err != nil {
-		return false, err
-	}
+	} else {
+		if err = p.config.ReferenceStore.AddTag(ref, imageID, true); err != nil {
+			return false, err
+		}
 
+		// Best effort store the digest, this could fail if the digest was
+		// already saved from a different tag.
+		digest, err := reference.WithDigest(ref, manifestDigest)
+		if err != nil {
+			logrus.Errorf("Failed to store digest reference %s: %s", manifestDigest, err.Error())
```

Comment:
This message should go with the call to `AddDigest` to log an error if one occurred. An error creating the reference would indicate that a reference or digest is bad, which _shouldn't_ be possible here, just returning the error should be fine.